server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -230,46 +231,71 @@ private class GetSnapshotsOperation {
 
-                    getSingleRepoSnapshotInfo(repoName, listeners.acquire((SnapshotsInRepo snapshotsInRepo) -> {
-                        allSnapshotInfos.add(snapshotsInRepo.snapshotInfos());
-                        remaining.addAndGet(snapshotsInRepo.remaining());
-                        totalCount.addAndGet(snapshotsInRepo.totalCount());
-                    }).delegateResponse((l, e) -> {
-                        if (isMultiRepoRequest && e instanceof ElasticsearchException elasticsearchException) {
-                            failuresByRepository.put(repoName, elasticsearchException);
-                            l.onResponse(SnapshotsInRepo.EMPTY);
-                        } else {
-                            l.onFailure(e);
+                            SubscribableListener
+
+                                .<RepositoryData>newForked(repositoryDataListener -> {
+                                    if (snapshotNamePredicate == SnapshotNamePredicate.MATCH_CURRENT_ONLY) {
+                                        repositoryDataListener.onResponse(null);
+                                    } else {
+                                        repositoriesService.repository(repoName).getRepositoryData(executor, repositoryDataListener);
+                                    }
+                                })
+
+                                .<SnapshotsInRepo>andThen((l, repositoryData) -> loadSnapshotInfos(repoName, repositoryData, l))
+
+                                .addListener(new DelegatingActionListener<>(listeners.acquire()) {
+                                    @Override
+                                    public void onResponse(SnapshotsInRepo snapshotsInRepo) {
+                                        allSnapshotInfos.add(snapshotsInRepo.snapshotInfos());
+                                        remaining.addAndGet(snapshotsInRepo.remaining());
+                                        totalCount.addAndGet(snapshotsInRepo.totalCount());
+                                        delegate.onResponse(null);
+                                    }
+
+                                    @Override
+                                    public void onFailure(Exception e) {
+                                        if (isMultiRepoRequest && e instanceof ElasticsearchException elasticsearchException) {
+                                            failuresByRepository.put(repoName, elasticsearchException);
+                                            delegate.onResponse(null);
+                                        } else {
+                                            delegate.onFailure(e);
+                                        }
+                                    }
+                                }, executor, threadPool.getThreadContext());
                         }
-                    }));
-                }
-            }
+                    }
+                })
+
+                .addListener(listener.map(ignored -> {
```

Comment:
This block of code executes _after_ the loop over the repositories, thought it'd be nicer to write it in execution order thusly.